eventually.go

```diff
@@ -0,0 +1,159 @@
+// since the intial successful check.
+//
+// Example calls:
+//
+// 		c.Assert(func() int64 {
+// 			return atomic.LoadInt64(&foo)
+// 		}, qt.EventuallyStable(qt.Equals), int64(1234))
+func EventuallyStable(c Checker) *EventuallyChecker {
+	eventuallyChecker := Eventually(c)
+	eventuallyChecker.WithStableStrategy(&retry.Strategy{
+		Delay:       100 * time.Millisecond,
+		MaxDuration: 150 * time.Millisecond,
+	})
+
+	return eventuallyChecker
+}
+
+// EventuallyChecker is a Checker that allows providing retry strategies to
+// retry the check over a period of time.
+// It also allows providing a stable retry strategy to run a stability check,
+// i.e. once the check is successful, it keeps checking that it stays that way.
+type EventuallyChecker struct {
+	checker             Checker
+	retryStrategy       *retry.Strategy
+	stableRetryStrategy *retry.Strategy
+}
+
+// WithStrategy allows specifying a custom retry strategy, specifying initial
+// delay, delay between attempts, maximum duration before timing out, etc.
+func (e *EventuallyChecker) WithStrategy(strategy *retry.Strategy) *EventuallyChecker {
+	e.retryStrategy = strategy
+	return e
+}
+
+// WithStableStrategy allows specifying a custom retry strategy for the
+// stability check. If not provided, no stability check will be run.
+func (e *EventuallyChecker) WithStableStrategy(strategy *retry.Strategy) *EventuallyChecker {
+	e.stableRetryStrategy = strategy
+	return e
+}
+
+// Check implements Checker.Check by calling the given got function repeatedly
+// and calling the underlying Checker with the returned value, according to the
+// retry Strategy, until the verification succeeds or the Strategy times out.
+// If the Strategy times out, the Check will fail.
+//
+// If an additional stable retry Strategy is provided, it also calls the
+// underlying Checker again after a succesfull check, according to the stable
+// retry Strategy, until either the Strategy times out or the verification
+// fails. If the stable Strategy times out, the Check will succeed.
+func (e *EventuallyChecker) Check(got interface{}, args []interface{}, note func(key string, value interface{})) error {
+	// Validate that the given got parameter is a function with no parameters
+	// and one return value.
+	f := reflect.ValueOf(got)
+	if f.Kind() != reflect.Func {
+		note("got", got)
+		return BadCheckf("first argument is not a function")
+	}
+	ftype := f.Type()
+	if ftype.NumIn() != 0 {
+		note("function", got)
+		return BadCheckf("cannot use a function receiving arguments")
+	}
+	if ftype.NumOut() != 1 {
+		note("function", got)
+		return BadCheckf("cannot use a function returning more than one value")
+	}
+
+	// Run the checker according to the retry strategy, succeeding on the first
+	// successful check.
+	for i := e.retryStrategy.Start(); ; {
+		got = f.Call(nil)[0].Interface()
+		err := e.checker.Check(got, args, note)
+		if err == nil {
+			break
+		}
+		if !i.Next(nil) {
+			note("got", got)
```

Comment:
Can't we trust that the underlying checker already notes the got argument?

Reply:
The problem is that the got argument is not noted by the checkers, it's appended to the output by the caller in `writeError` according to the parameter passed to the Eventually checker, which is a function.  So the output will look like this:
```
got: func () int {...}
want: 42
```

We need overwrite that and that's why I added it as a note and then changed the logic slightly in `writeError` to skip any provided argument if they are already in the notes.